prow/plugins/invalidcommitmsg/invalidcommitmsg.go

```diff
@@ -0,0 +1,174 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package invalidcommitmsg adds the "do-not-merge/invalid-commit-message"
+// label on PRs containing commit messages with @mentions or
+// keywords that can automatically close issues.
+package invalidcommitmsg
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/plugins/dco"
+)
+
+const (
+	pluginName            = "invalidcommitmsg"
+	invalidCommitMsgLabel = "do-not-merge/invalid-commit-message"
+	commentBody           = `[Keywords](https://help.github.com/articles/closing-issues-using-keywords) which can automatically close issues and at(@) mentions are not allowed in commit messages.
+
+**The list of commits with invalid commit messages**:
+
+%s
+
+<details>
+
+%s
+</details>
+`
+)
+
+var (
+	closeIssueRegex = regexp.MustCompile(`((?i)(clos(?:e[sd]?))|(fix(?:(es|ed)?))|(resolv(?:e[sd]?)))[\s:]+(\w+/\w+)?#(\d+)`)
+	atMentionRegex  = regexp.MustCompile(`\s([@][\w_-]+)`)
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+			Description: "The invalidcommitmsg plugin applies the '" + invalidCommitMsgLabel + "' label to pull requests whose commit messages contain @ mentions or keywords which can automatically close issues.",
+		},
+		nil
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	CreateComment(owner, repo string, number int, comment string) error
+	ListPRCommits(org, repo string, number int) ([]github.RepositoryCommit, error)
+}
+
+type commentPruner interface {
+	PruneComments(shouldPrune func(github.IssueComment) bool)
+}
+
+func handlePullRequest(pc plugins.Agent, pr github.PullRequestEvent) error {
+	cp, err := pc.CommentPruner()
+	if err != nil {
+		return err
+	}
+	return handle(pc.GitHubClient, pc.Logger, pr, cp)
+}
+
+func handle(gc githubClient, log *logrus.Entry, pr github.PullRequestEvent, cp commentPruner) error {
+	// Only consider actions indicating that the code diffs may have changed.
+	if !hasPRChanged(pr) {
+		return nil
+	}
+
+	shouldComment := pr.Action == github.PullRequestActionSynchronize ||
+		pr.Action == github.PullRequestActionOpened
```

Comment:
Please add a (code) comment to indicate why we only want to comment (on GH) for these specific actions.

Reply:
Thinking more about this, I removed the condition because "reopen" actions sometimes make it easy to retrigger bot automation in the case of a dropped webhook, etc.